### ISSUE TITLE
Missing gflags-targets.cmake, package registry entry, misc

### ISSUE THREAD

**Issue by jiawen** [gflags/gflags#174]
(Disclaimer: I'm only a CMake journeyman so this could be a CMake issue. Please correct me if I'm wrong.)
## TL;DR

`gflags-config.cmake` seems to export `gflags_INCLUDE_DIR` and `gflags_LIBRARIES`, but not something like `gflags_LIBRARY_DIR`. As a result, using the default instructions, linking fails with `error LNK1181: cannot open input file 'gflags.lib'`.

When built statically, the CMake config also fails to inform the generated project that it depends on `Shlwapi.lib` and linking fails with: "unresolved external symbol __imp_PathMatchSpecA ..."

By manually adding the library search path to the Visual Studio project and adding `Shlwapi.lib` as a dependency, my project builds fine.
## Details

I cloned and built gflags on Windows and successfully built it with VS2013:
- cd c:\tmp\gflags-2.1.2
- mkdir build
- cd build
- cmake-gui ..
- <open .sln and build with VS2013>

Building the INSTALL target on Windows puts everything under C:\Program Files\libraries\gflags, which has as children:
- CMake
- Include
- Lib

In my own project, I followed the directions as on https://gflags.github.io/gflags/, except to find `gflags.h`, I had to add `include_directories( ${gflags_INCLUDE_DIR} )` to tell the compiler where to find the header. `gflags_INCLUDE_DIR` resolves to `C:\Program Files\libraries\gflags\Include` and `gflags_LIBRARIES` resolves to `gflags.lib`. However, the compiler can't find `gflags.lib` because there isn't a `gflags_LIBRARY_DIR` or something to that effect which resolves to `C:\Program Files\libraries\gflags\Lib`.

**Comment by schuhschuh** [comment]
Thanks for the detailed report.

A variable such as `gflags_LIBRARY_DIR` should not be required. This was common in the beginnings of CMake, but nowadays import targets with the full library path attached as property are preferred. This path is specified in the `gflags-targets.cmake` file. The `gflags_LIBRARIES` variable should simply contain the string `gflags` and is provided only to be "compatible" with other standard CMake Find modules. You could also just use `gflags`, which is the name of the respective CMake import target, not a file name such as `gflags.lib`. Alternatively, you can use one of the specific exported targets found in `gflags-targets.cmake` in the `target_link_libraries` command. Moreover, also the `gflags_INCLUDE_DIR` should not have to be specified explicitly because the `gflags` import target has this path as transitive include dependency property set.

You really should only need the CMake code found at https://gflags.github.io/gflags/#cmake.

Can you provide some more information to see why this doesn't seem to be sufficient in your case? For example:

- CMake version?
- Can you find the following in `CMake/gflags-targets.cmake` file and are the paths correct for your installation?

```cmake
# Create imported target gflags_nothreads_static
add_library(gflags_nothreads_static STATIC IMPORTED)

set_target_properties(gflags_nothreads_static PROPERTIES
  INTERFACE_INCLUDE_DIRECTORIES "/Users/as12312/Software/gflags/xcode/include"
)

# [...]

# Import target "gflags_nothreads_static" for configuration "Release"
set_property(TARGET gflags_nothreads_static APPEND PROPERTY IMPORTED_CONFIGURATIONS RELEASE)
set_target_properties(gflags_nothreads_static PROPERTIES
  IMPORTED_LINK_INTERFACE_LANGUAGES_RELEASE "CXX"
  IMPORTED_LOCATION_RELEASE "/Users/as12312/Software/gflags/xcode/lib/Release/libgflags_nothreads.a"
  )
```

- Does it work if you use `target_link_libraries(foo gflags_nothreads_static)` instead of `gflags_LIBRARIES` or `gflags` import target, respectively? I am asking because in the `gflags-config.cmake` file, there is some CMake code to mimic a `gflags` ALIAS import target which is currently not possible with CMake yet. The actual import target names are `gflags_static`, `gflags_shared`, `gflags_nothreads_static`, and `gflags_nothreads_shared`. The `gflags` import target is intended to be an alias for either one of these as the "default" target to be used. If your project builds fine (i.e., the library file is found) using `gflags_nothreads_static` instead of `gflags`, the issue may be with this CMake code in `gflags-config.cmake`.

Before trying, please remove the `include_directories(${gflags_INCLUDE_DIR})` again from your project. While this is not wrong, I would like it to be not needed. You really should be fine with just the `target_link_libraries` command.


> When built statically, the CMake config also fails to inform the generated project that it depends on Shlwapi.lib and linking fails with: "unresolved external symbol __imp_PathMatchSpecA ..."

Thanks, I'll take note to add `shlwapi.lib` as link dependency of the static import targets.

**Comment by schuhschuh** [comment]
BTW You could also use [the example project](https://github.com/gflags/example) to test the CMake configuration of your gflags installation. The CMakeLists.txt of this project basically comes down to only the following:

```cmake
find_package(gflags REQUIRED)
# add_subdirectory(foo)
add_executable(foo foo.cc)
target_link_libraries(foo gflags)
```

**Comment by jiawen** [comment]
Thanks for the super fast and detailed reply!

The example project: this builds out of the box and works great. However, it uses gflags as a Git submodule and is compiled in-tree. I'm trying to have my project use the one installed on the system, which I guess is pretty weird on Windows.

Re: cmake version: 3.5.1 (on Windows 8.1 x64).

In the course of going through your suggestions, I figured out what the problem was. I was mistaken when I said I cloned gflags from the source repo - instead, I got it by downloading the 2.1.2 release .zip. For some reason, doing a CMake build and install from the release results in the following directory tree:

Re: `${CMAKE_INSTALL_PREFIX}/CMake/gflags-targets.cmake`
This file doesn't exist! The complete contents of my install directory:
`CMAKE_INSTALL_PREFIX`: C:\Program Files\gflags
* CMake
  * gflags-config-version.cmake
  * gflags-config.cmake
  * gflags-export-release.cmake
  * gflags-export.cmake
* Include
  * gflags
    * gflags.h
    * gflags_completions.h
    * gflags_declare.h
    * gflags_gflags.h
* Lib
  * gflags.lib
  * gflags_nothreads.lib

Compare to a build from cloning the repo (differences in **bold**):
* CMake
  * gflags-config-version.cmake
  * gflags-config.cmake
  * **gflags-targets.cmake**
  * **gflags-targets-debug.cmake**
  * **gflags-targets-release.cmake**
* Include
  * gflags
    * gflags.h
    * gflags_completions.h
    * gflags_declare.h
    * gflags_gflags.h
* Lib
  * **gflags_nothreads_static.lib**
  * **gflags_static.lib**

Any idea what's going on there?

**Comment by jiawen** [comment]
More annoyingly, and this might be a CMake bug rather than how gflags is using it, when I run the INSTALL target, it adds the *build directory* to the registry and *not* where it gets installed.

I.e., even though `CMAKE_INSTALL_PREFIX` is set to the default value of `C:\Program Files\gflags`, it creates a registry key under `HKEY_CURRENT_USER\Software\Kitware\CMake\Packages\gflags`: `78247c9cf0da8a791e30633c4a8845cf` --> `c:/tmp/gflags/bld`.

(I cloned into c:\tmp\gflags. Then did mkdir bld; cd bld; cmake-gui ..).

**Comment by jiawen** [comment]
Btw, you don't have to fix the dependency on `shlwapi.lib`, when building and installing out of the clone, the -targets.cmake file correctly adds it as a dependency.

**Comment by schuhschuh** [comment]
> The example project: this builds out of the box and works great. However, it uses gflags as a Git submodule and is compiled in-tree. I'm trying to have my project use the one installed on the system, which I guess is pretty weird on Windows.

Good! You can also test it by using your own installation of the gflags library. Therefore, simply do not initialize/update the Git submodule as explained in the README. As you have done so already, just run the following Git command before you re-execute CMake (possibly in new clean build directory):

```
git submodule deinit gflags
```

### Issue 1

> Re: ${CMAKE_INSTALL_PREFIX}/CMake/gflags-targets.cmake
This file doesn't exist!

Surprising that you don't get a CMake error then when `gflags-config.cmake` is included by the `find_package(gflags)` command. Because this file includes `gflags-targets.cmake` and this `include` is not optional. In any case, if this file is missing, it explains that things aren't working.

[This is the line responsible for generating/installing this file](https://github.com/gflags/gflags/blob/f4eace133187e0a101a6d6d71c55592b572de189/CMakeLists.txt#L515).

The mistake is that I used `FILE` instead of `RENAME` and thus it uses the default name `gflags-export.cmake` instead of `gflags-targets.cmake`.

- [ ] Fix `install(EXPORT ...)` command, specify file name with `RENAME`.

### Issue 2

The library file names on Windows differ from Unix systems. On Unix, we can distinguish static from dynamic libraries based on the file name extension alone. On Windows, we need to have different names for the import library `.lib` to not confuse/overwrite it with a static library which has the same file name extension as an import library.

Are you sure you set `BUILD_SHARED_LIBS` to OFF (and `BUILD_STATIC_LIBS` to ON, but the default if both are OFF) in both cases? But then I would expect the corresponding `.dll` files in the `Lib` directory as well... not sure why the library file name is not as expected, i.e., `gflags_static.lib` and `gflags_nothreads_static.lib`. See [here](https://github.com/gflags/gflags/blob/f4eace133187e0a101a6d6d71c55592b572de189/CMakeLists.txt#L448).

- [ ] Check file names of installed static libraries (on Windows).

### Issue 3

> More annoyingly, and this might be a CMake bug rather than how gflags is using it, when I run the INSTALL target, it adds the build directory to the registry and not where it gets installed.

"The User Package Registry is stored in a per-user location. The export(PACKAGE) command may be used to register a project build tree in the user package registry. CMake currently provides no interface to add install trees to the user package registry. Installers must be manually taught to register their packages if desired." (https://cmake.org/Wiki/CMake/Tutorials/Package_Registry)

In my [CMake BASIS](http://cmake-basis.github.io/) project, I actually wrote a function to register the install tree instead. See [here](https://github.com/cmake-basis/BASIS/blob/65930fd83df4c5e69f990a5fab6c9e78a3c28835/src/cmake/modules/InstallationTools.cmake#L404).

- [ ] Update CMake package registry entry to change path during installation

**Comment by jiawen** [comment]
Wow, you're quite the CMake master!

### Re: gflags-example.
It works great once I change `target_link_libraries(foo gflags)` to `target_link_libraries(foo gflags_static)` or `target_link_libraries(foo gflags_shared)`. Related directly to issue 2.

### Re: Issue 2
I just checked, doing a full delete of the CMake cache and all generated projects between each run.

With the defaults (both `BUILD_SHARED_LIBS = OFF` and `BUILD_STATIC_LIBS = OFF` ), CMake builds and installs `gflags_static.lib` and `gflags_nothreads_static.lib` However, in `gflags-targets.cmake`, the corresponding targets are `gflags_static`, and `gflags_nothreads_static`.

The same happens if `BUILD_SHARED_LIBS = OFF` and `BUILD_STATIC_LIBS = ON`, as expected.

When `BUILD_SHARED_LIBS = ON and BUILD_STATIC_LIBS = OFF`, it builds and installs `gflags.dll`, `gflags.lib`, gflags_nothreads.dll`, and `gflags_nothreads.lib`. The corresponding targets are `gflags_shared`, `gflags_nothreads_shared`.

When `BUILD_SHARED_LIBS = ON and BUILD_STATIC_LIBS = ON`, it builds and installs `gflags.dll`, `gflags.lib`, gflags_nothreads.dll`, and `gflags_nothreads.lib`. The corresponding targets are `gflags_shared`, `gflags_nothreads_shared`, `gflags_static`, and `gflags_nothreads_static`.

__In all cases, there is no target just called `gflags`.__

### Re: Issue 3
Awesome, thanks!

**Comment by schuhschuh** [comment]
> I just checked, doing a full delete of the CMake cache and all generated projects between each run.

Perfect, thanks for trying all combinations.

For the static builds, everything looks fine. However, I am curious why in case of `BUILD_SHARED_LIBS` and `BUILD_STATIC_LIBS` your installation seems to be missing the `gflags_static.lib` and `gflags_nothreads_static.lib` files?

> In all cases, there is no target just called gflags.

CMake has so far no ability to define an imported [ALIAS library target](https://cmake.org/cmake/help/v3.6/command/add_library.html#alias-libraries). This feature you can currently only use within your own project. Hence, I add the `gflags` target myself in the `gflags-config.cmake` file and copy the attributes from either one of the targets in `gflags-targets.cmake` as exported by CMake. The respective default target is chosen [here](https://github.com/gflags/gflags/blob/f4eace133187e0a101a6d6d71c55592b572de189/cmake/config.cmake.in#L74) and the `gflags` (i.e., `@PACKAGE_NAME@`) is added [here](https://github.com/gflags/gflags/blob/f4eace133187e0a101a6d6d71c55592b572de189/cmake/config.cmake.in#L97).

**Comment by jiawen** [comment]
Related even more obnoxious Windows issue: debug vs release C++ runtime libraries and _ITERATOR_DEBUG_LEVEL.

I'm working on my project which depends on gflags and during development, I need to run both the debug and release builds. If gflags is built in `SHARED` mode, everything works fine. However, if built with `STATIC`, then I have to link again a matching version of gflags since gflags depends on the C++ runtime library and (implicitly) sets the _ITERATOR_DEBUG_LEVEL value to 0 or 2. I.e., the `DEBUG` version of my app has to link to the `DEBUG` version of gflags and similarly for the `RELEASE` versions.

I'm not a CMake expert, but looking at the generated gflags-targets.cmake file, it looks like CMake is smart enough to point a target at different static libraries for different configurations.

    # Import target "gflags_static" for configuration "Debug"
    set_property(TARGET gflags_static APPEND PROPERTY IMPORTED_CONFIGURATIONS DEBUG)
    set_target_properties(gflags_static PROPERTIES
      IMPORTED_LINK_INTERFACE_LANGUAGES_DEBUG "CXX"
      IMPORTED_LOCATION_DEBUG "C:/tmp/gflags/bld/lib/Debug/gflags_static.lib"
    )

    # Import target "gflags_static" for configuration "Release"
    set_property(TARGET gflags_static APPEND PROPERTY IMPORTED_CONFIGURATIONS RELEASE)
    set_target_properties(gflags_static PROPERTIES
      IMPORTED_LINK_INTERFACE_LANGUAGES_RELEASE "CXX"
      IMPORTED_LOCATION_RELEASE "C:/tmp/gflags/bld/lib/Release/gflags_static.lib"
    )

The convention is to append a `d` at the end of the lib file, i.e., `gflags_staticd.lib`. And then we just have to point the `DEBUG` configuration of the target to that file. The magic CMake command seems to be:

    set( CMAKE_DEBUG_POSTFIX d )

**Comment by jiawen** [comment]
Oops, copy paste fail and there are too many libs, dlls, and targets :). I should have checked the preview first. You're right, with both checkboxes ON, it generates all 4 combinations, `gflags_static`, `gflags_nothreads_static`, `gflags_shared`, and `gflags_nothreads_shared`.

**Comment by schuhschuh** [comment]
Yes, I am aware of the Debug library file prefix. However, I would not really expect a library with debug symbols to be part of a package installation. If you are a developer and need to debug your code, you should compile the `gflags` dependency yourself from sources anyway instead of using a binary package (e.g., Ubuntu/Debian package installation, Windows installer...). In this case, you should skip the `install` step in simply use the build tree of gflags directly as your `gflags_DIR` when configuring your dependent project.

P.S.: For other issues, I think it would be better if we would discuss separate issues in separate threads... makes tracking required changes easier :-).

**Comment by schuhschuh** [comment]
> set( CMAKE_DEBUG_POSTFIX d )

Regarding the debug postfix, if desired, this setting can be modified on the `cmake` command line using the `-D` option or in the CMake GUI.

**Comment by schuhschuh** [comment]
Re: Issue 1: I couldn't find anything wrong with using `install(EXPORT)`. To avoid issues relating to the renaming of the file, however, I introduced the internal variable `EXPORT_NAME` and avoid the use of the `FILE` option to rename the generated file upon installation. See #184.

**Comment by schuhschuh** [comment]
Re: Issue 3: Added options `REGISTER_BUILD_DIR` and `REGISTER_INSTALL_PREFIX` where the latter is the default now when the gflags project is not a sub-project of another project. See #185.

**Comment by schuhschuh** [comment]
> The respective default target is chosen here and the gflags (i.e., @PACKAGE_NAME@) target is added here.

The imported `gflags` target was missing the `IMPORTED_IMPLIB_<CONFIG>` property, i.e., the path of the import library. Fixed by #186.
